File: Little-book-of-semaphores/3.6/Barrier.java
import java.util.concurrent.Semaphore;

public class Barrier {
    public static void main(String[] args) {
        Semaphore mutex = new Semaphore(1);
        Semaphore sem1 = new Semaphore(0);
        Semaphore sem2 = new Semaphore(1);
        final int[] count = {0};
        int barrierSize = 4;

        Runnable barrier = () -> {
            try {
                mutex.acquire();
                count[0]++;
                mutex.release();

                if (count[0] == barrierSize) {
                    sem1.release();
                }

                sem1.acquire();
                sem1.release();

                System.out.println(Thread.currentThread().getName() + " passou pela barreira");

            } catch (Exception e) {
                e.printStackTrace();
            }
        };

        int totalThreads = 4;
        for (int i = 0; i < totalThreads; i++) {
            new Thread(barrier, "Thread-" + (i + 1)).start();
        }
    }
}

File: Little-book-of-semaphores/3.6/Barrier.py
from threading import Thread
from threading import Lock
from threading import Barrier

mutex = Lock()
barrier = Barrier(4)

count = 0

def sum():
    global count
    with mutex:
        count = count + 1
        print(count)
    barrier.wait()
    print(f'Finalizei {count}')

threads = [Thread(target=sum) for _ in range(4)]

for thread in threads:
    thread.start()

for thread in threads:
    thread.join()